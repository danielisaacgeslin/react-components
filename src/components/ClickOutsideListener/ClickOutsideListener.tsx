import * as React from 'react';

export interface IClickOutsideListenerProps {
  onClickOutside: () => void;
  children?: any;
  document?: Document | Element;
  insideRef?: Element;
}

export default class ClickOutsideListener extends React.PureComponent<IClickOutsideListenerProps> {
  public static defaultProps: IClickOutsideListenerProps = {
    document: window.document,
    insideRef: null,
    onClickOutside: /* istanbul ignore next */ () => null
  };

  public ref = React.createRef<HTMLDivElement>();

  public componentDidMount() {
    const { document } = this.props;
    document.addEventListener('mousedown', this.handleClick);
  }

  public componentWillUnmount() {
    const { document } = this.props;
    document.removeEventListener('mousedown', this.handleClick);
  }

  public handleClick = event => {
    const { onClickOutside } = this.props;
    const insideRef = this.props.insideRef || this.ref.current;
    if (insideRef && !insideRef.contains(event.target)) onClickOutside();
  };

  public render() {
    const { children } = this.props;
    return (
      <div key="click-outside-listener" ref={this.ref}>
        {children}
      </div>
    );
  }
}
