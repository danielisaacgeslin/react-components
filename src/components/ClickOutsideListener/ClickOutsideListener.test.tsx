import * as React from 'react';
import { create } from 'react-test-renderer';
import { mount } from 'enzyme';

import ClickOutsideListener from './ClickOutsideListener';

describe('ClickOutsideListener Component', () => {
  global.console.error = () => null;
  let wrapper;
  let props;
  let onClickOutside;
  let addEventListener;
  let removeEventListener;
  let triggerEventOutside;
  let triggerEventInside;
  let registry;

  beforeEach(() => {
    registry = {};
    onClickOutside = jest.fn();
    removeEventListener = jest.fn();

    addEventListener = (eventName, cb) => (registry[eventName] = event => cb(event));

    props = { onClickOutside, document: { addEventListener, removeEventListener } };

    wrapper = mount(
      <div key="tst">
        <span className="outside">outside</span>
        <ClickOutsideListener {...props}>
          <span className="inside">inside</span>
        </ClickOutsideListener>
      </div>
    );

    triggerEventOutside = eventName =>
      registry[eventName]({
        target: wrapper
          .find('.outside')
          .first()
          .getDOMNode()
      });
    triggerEventInside = eventName =>
      registry[eventName]({
        target: wrapper
          .find('.inside')
          .first()
          .getDOMNode()
      });
  });

  it('should render children', () => {
    expect(create(wrapper).toJSON()).toMatchSnapshot();
  });

  it('should call onClickOutside when click happens OUTSIDE', () => {
    triggerEventOutside('mousedown');
    expect(onClickOutside).toBeCalled();
  });

  it('should NOT call onClickOutside when click happens INSIDE', () => {
    triggerEventInside('mousedown');
    expect(onClickOutside).not.toBeCalled();
  });

  it('should call removeEventListener', () => {
    wrapper.unmount();
    expect(removeEventListener).toBeCalled();
  });
});
