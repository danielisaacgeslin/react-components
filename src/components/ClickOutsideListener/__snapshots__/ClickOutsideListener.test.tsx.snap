// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickOutsideListener Component should render children 1`] = `
<div>
  <span
    className="outside"
  >
    outside
  </span>
  <div>
    <span
      className="inside"
    >
      inside
    </span>
  </div>
</div>
`;
